File: problems/Java/13. Roman to Integer.java
class Solution {
    public int romanToInt(String s) {
        int res = 0;
        int prev = 0;
        int val = 0;
        
        for (int i = s.length() - 1; i >= 0; i--) {
            switch (s.charAt(i)) {
                case 'I': val = 1; break;
                case 'V': val = 5; break;
                case 'X': val = 10; break;
                case 'L': val = 50; break;
                case 'C': val = 100; break;
                case 'D': val = 500; break;
                case 'M': val = 1000; break;
            }
            
            if (val < prev) {
                res -= val;
            } else {
                res += val;
            }
            prev = val;
        }
        return res;
    }
}

File: problems/C++/13. Roman to Integer.cpp
class Solution {
public:
    int romanToInt(string s) {
        unordered_map<char, int> rmap = {
            {'I', 1}, 
            {'V', 5}, 
            {'X', 10}, 
            {'L', 50}, 
            {'C', 100}, 
            {'D', 500}, 
            {'M', 1000}
        };
        
        int res = 0;
        int prev = 0;
        
        for (int i = s.length() - 1; i >= 0; --i) {
            int val = rmap[s[i]];
            if (val < prev) {
                res -= val;
            } else {
                res += val;
            }
            prev = val;
        }
        return res;
    }
};
